chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+	}
+
+	txHash, err := ethTx.TxHash()
+	if err != nil {
+		return xerrors.Errorf("error hashing transaction: %w", err)
+	}
+
+	return si.indexEthTxHash(ctx, tx, txHash, msg.Cid())
+}
+
+func (si *SqliteIndexer) indexEthTxHash(ctx context.Context, tx *sql.Tx, txHash ethtypes.EthHash, msgCid cid.Cid) error {
+	insertEthTxHashStmt := tx.Stmt(si.insertEthTxHashStmt)
+	_, err := insertEthTxHashStmt.ExecContext(ctx, txHash.String(), msgCid.Bytes())
+	if err != nil {
+		return xerrors.Errorf("failed to index eth tx hash: %w", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) Apply(ctx context.Context, from, to *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	// We're moving the chain ahead from the `from` tipset to the `to` tipset
+	// Height(to) > Height(from)
+	err := withTx(ctx, si.db, func(tx *sql.Tx) error {
+		// index the `to` tipset first as we only need to index the tipsets and messages for it
+		if err := si.indexTipset(ctx, tx, to); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		// index the `from` tipset just in case it's not indexed
+		if err := si.indexTipset(ctx, tx, from); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return xerrors.Errorf("error applying tipset: %w", err)
+	}
+
+	si.notifyUpdateSubs()
+
+	return nil
+}
+
+func (si *SqliteIndexer) Revert(ctx context.Context, from, to *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	// We're reverting the chain from the tipset at `from` to the tipset at `to`.
+	// Height(to) < Height(from)
+
+	revertTsKeyCid, err := toTipsetKeyCidBytes(from)
+	if err != nil {
+		return xerrors.Errorf("error getting tipset key cid: %w", err)
+	}
+
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		if _, err := tx.Stmt(si.revertTipsetStmt).ExecContext(ctx, revertTsKeyCid); err != nil {
+			return xerrors.Errorf("error marking tipset as reverted: %w", err)
+		}
+
+		// index the `to` tipset as it has now been applied -> simply for redundancy
+		if err := si.indexTipset(ctx, tx, to); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("error during revert transaction: %w", err)
+	}
+
+	si.notifyUpdateSubs()
+
+	return nil
+}
+
+func (si *SqliteIndexer) indexTipset(ctx context.Context, tx *sql.Tx, ts *types.TipSet) error {
+	tsKeyCidBytes, err := toTipsetKeyCidBytes(ts)
+	if err != nil {
+		return xerrors.Errorf("error computing tipset cid: %w", err)
+	}
+
+	restored, err := si.restoreTipsetIfExists(ctx, tx, tsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("error restoring tipset: %w", err)
+	}
+	if restored {
+		return nil
+	}
+
+	height := ts.Height()
+	insertTipsetMsgStmt := tx.Stmt(si.insertTipsetMessageStmt)
+
+	msgs, err := si.cs.MessagesForTipset(ctx, ts)
```

Comment:
You're dobling up on calls to `MessagesForBlock` because `MessagesForTipset` calls it internally. There is a cache (`mmCache`) but this still looks pretty inefficient. Could you investigate minimising duplication? The problem I can see is that `MessagesForTipset` does some de-duplication on sender & nonce but `MessagesForBlock` is just what it says on the label. What do we need the strict de-duplicated list of messages in `tipset_messages` or can we just put them all in there? Could we consider doing the de-duplication ourselves (although that may mean exposing a utility function that can be shared with the ChainStore).

Reply:
@rvagg But we only call this once per tipset i.e. every 30 seconds. Do you still think it is a problem ? Can you elaborate on why you think this should be fixed ?

Reply:
Because everything we add to tipset processing adds to the 30s overhead. Badger isn't awesome, block loads aren't cheap, the less we have to touch it the best and if there's dedupe potential we should look at it.

<img width="408" alt="Screenshot 2024-09-12 at 9 53 07 PM" src="https://github.com/user-attachments/assets/94df10c8-a875-4345-bf94-1951ccfd89e3">

maybe this is a stretch goal but it does bother me that `MessagesForTipset` calls `MessagesForBlock` internally, yet we're also calling `MessagesForBlock` ourselves for each block, the only thing stopping us from doing it the other way around is the `selectMsg` closure.

<details>
<summary>diff that exposes that</summary>

```diff
diff --git a/chain/store/messages.go b/chain/store/messages.go
index 60a1eaa60..8b3871242 100644
--- a/chain/store/messages.go
+++ b/chain/store/messages.go
@@ -103,21 +103,20 @@ type BlockMessages struct {
        SecpkMessages []types.ChainMsg
 }
 
-func (cs *ChainStore) BlockMsgsForTipset(ctx context.Context, ts *types.TipSet) ([]BlockMessages, error) {
-       // returned BlockMessages match block order in tipset
-
+// MessageSelector returns a function that can be used to filter messages for a block based on their nonce.
+func MessageSelector(stateBlockstore bstore.Blockstore, ts *types.TipSet) (func(m *types.Message) (bool, error), error) {
        applied := make(map[address.Address]uint64)
 
-       cst := cbor.NewCborStore(cs.stateBlockstore)
+       cst := cbor.NewCborStore(stateBlockstore)
        st, err := state.LoadStateTree(cst, ts.Blocks()[0].ParentStateRoot)
        if err != nil {
                return nil, xerrors.Errorf("failed to load state tree at tipset %s: %w", ts, err)
        }
 
        useIds := false
-       selectMsg := func(m *types.Message) (bool, error) {
-               var sender address.Address
-               if ts.Height() >= buildconstants.UpgradeHyperdriveHeight {
+       if ts.Height() >= buildconstants.UpgradeHyperdriveHeight {
+               return func(m *types.Message) (bool, error) {
+                       var sender address.Address
                        if useIds {
                                sender, err = st.LookupIDAddress(m.From)
                                if err != nil {
@@ -144,24 +143,42 @@ func (cs *ChainStore) BlockMsgsForTipset(ctx context.Context, ts *types.TipSet)
                                        }
                                }
                        }
-               } else {
-                       sender = m.From
-               }
 
+                       // The first match for a sender is guaranteed to have correct nonce -- the block isn't valid otherwise
+                       if _, ok := applied[sender]; !ok {
+                               applied[sender] = m.Nonce
+                       }
+                       if applied[sender] != m.Nonce {
+                               return false, nil
+                       }
+                       applied[sender]++
+
+                       return true, nil
+               }, nil
+       }
+
+       // else < buildconstants.UpgradeHyperdriveHeight
+       return func(m *types.Message) (bool, error) {
                // The first match for a sender is guaranteed to have correct nonce -- the block isn't valid otherwise
-               if _, ok := applied[sender]; !ok {
-                       applied[sender] = m.Nonce
+               if _, ok := applied[m.From]; !ok {
+                       applied[m.From] = m.Nonce
                }
-
-               if applied[sender] != m.Nonce {
+               if applied[m.From] != m.Nonce {
                        return false, nil
                }
-
-               applied[sender]++
+               applied[m.From]++
 
                return true, nil
+       }, nil
+}
+
+func (cs *ChainStore) BlockMsgsForTipset(ctx context.Context, ts *types.TipSet) ([]BlockMessages, error) {
+       selectMsg, err := MessageSelector(cs.stateBlockstore, ts)
+       if err != nil {
+               return nil, err
        }
 
+       // returned BlockMessages match block order in tipset
        var out []BlockMessages
        for _, b := range ts.Blocks() {
```
</details>

^ it would be quite a change to integrate message selecting and order into here so I'm not going to push the point and just hope the cache is enough

Reply:
Okay, let's park this for a follow up PR if it becomes a problem. The only thing I'll add here is that we're not adding any additional cost here that did not already exist because the existing `MsgIndex` and `EventIndex` that we're getting rid of also rely on this method.

Reply:
👍 good point, reasonable to look at this later since it's not an easy solve now